resources/lib/player/players.py

```diff
@@ -145,17 +146,27 @@ def _set_external_ids(self, tmdb_type, tmdb_id, season, episode, details, requir
         return set_detailed_item(tmdb_type, tmdb_id, season, episode, details=details) or {}
 
     def _set_provider_priority(self):
-        try:  # Check if players are listed in providers and bump priority if found
-            providers = self.details.infoproperties['providers']
-            providers = providers.split(' / ')
-            for k, v in self.players.items():
-                if 'provider' not in v or v['provider'] not in providers:
-                    v['priority'] = v.get('priority', PLAYERS_PRIORITY) + 100  # Increase priority baseline by 100 to prevent other players displaying above providers
-                    continue
-                v['priority'] = providers.index(v['provider']) + 1  # Add 1 because sorted() puts 0 index last
-                v['is_provider'] = True
-        except (KeyError, AttributeError):
-            pass
+        try:
+            providers = self.details.infoproperties.get('providers', '')
+        except AttributeError:
+            providers = ''
+        providers = tuple(providers.split(' / '))
+
+        def _set_priority(item, _providers=providers):
+            file, player = item
+            player_provider = player.get('provider')
+            if player_provider in _providers:
+                priority = _providers.index(player_provider) + 1  # Add 1 because sorted() puts 0 index last
+                player['is_provider'] = True
+            else:
+                if player.get('is_provider', True):
+                    priority = player.get('priority', PLAYERS_PRIORITY) + 100  # Increase priority baseline by 100 to prevent other players displaying above providers
+                player['is_provider'] = False    
+            player['priority'] = priority
+            return priority, u_normalize('NFKD', player.get('name', '\uFFFF')).lower()
+
+        players = sorted(self.players.items(), key=_set_priority)
```

Comment:
The idea was to minimise the number of times the players were sorted and looped through.

If there were providers, the players were looped through. The players were then being sorted and then looped through again to get dialog entries.

And in order to select a player by plugin name or provider, via the player parameter, while also respecting player priority and mode, I needed to sort and loop through the players again.

The aim of this PR was to perform the selection by name/provider via the player parameter, and for that I need to update the priorities, sort by priority, then loop through the result. The fact that the provider priority is being updated was secondary to that, because the sorting with priority as the sort key was required anyway, which is why I removed the KeyError handling.

Perhaps _set_provider_priority can be renamed to _get_prioritised_players? It would save having to iterate over players twice in a row if providers were present.

Reply:
[9c4ffee](https://github.com/jurialmunkey/plugin.video.themoviedb.helper/pull/965/commits/9c4ffeecc14336c69e20cadf1e2338332facf99d) renames _set_provider_priority to _get_prioritised_players, but if you prefer, I can split this up into two different loops - setting priority in _set_provider_priority, and sorting by priority in _get_prioritised_players.

Reply:
Oh, of course, I see what's happening here now! You've skipped the first for loop by making it part of sorted() key function.  

I didn't pick that up at first, so was still thinking there would be two loops either way. Hadn't thought of using sorted() as a substitute for loop before. That's actually really clever.

In that case, I'm definitely happy with that change. Cutting out an unnecessary loop beats slightly easier to read code.

Thanks for persisting with me here! Your code is definitely 100% an improvement and was just my idiot brain misunderstanding exactly what it was doing.

Reply:
All good. Trying to review someone else's code isn't always easy, and you need to be comfortable with the changes. Thanks for taking the time to review.